src/main/java/fi/aalto/cs/apluscourses/presentation/MainViewModel.java

```diff
@@ -118,4 +121,17 @@ public ExercisesTreeViewModel getExercises() {
   public Options getExerciseFilterOptions() {
     return exerciseFilterOptions;
   }
+
+  /**
+   * <p>Returns the name of the Card to be shown based on whether a Course object exists
+   * and is not null.</p>
+   */
+  public String getCardToBeShown() {
+    if (Optional.ofNullable(courseViewModel.get())
+            .map(BaseViewModel::getModel)
+            .orElse(null) != null) {
+      return "TreeCard";
+    }
+    return "LabelCard";
+  }
```

Comment:
Nit: the cards are sort of an implementation detail of the UI, so it might make more sense to have methods like `isEmptyTextVisible` or `isListVisible`, but this isn't very important though

Reply:
A good point in that the strings don't have to refer to a card layout, but can instead be seen as just some descriptions of what is shown.

Strings are quite brittle for this though, since you rely on magic constants that both the view model and view have to understand, and the method has to return a specific subset of strings, even though there are infinitely many strings.

In this case it works quite neatly, but that is only because the view uses a card layout (implementation detail)

Reply:
It's just a Java way to use constants, either integers or strings, to pass "enum" values (actual enums being not that nice to work with). Using either string or int valued constants would make sense in this class's interface, strings perhaps a bit more because there's no order between them.

Similarly, I think that the designers of Swing made a good choice when having CardLayout use string keys (instead of numeric indexing). I guess they had this kind of usage in mind.

About those boolean methods: when thinking those two methods as one `MainViewModel -> Boolean * Boolean` valued function, we see that only values `(true, false)` and `(false, true)` make sense. `(true, true)` (i.e. show both) and `(false, false)` (show neither) are "nonsensical". ;)

Reply:
Yeah good point that half of the boolean pairs are nonsensical as well :+1: (I also considered enums because of that, but as you note, they aren't that nice to work with)

CardLayout using string vs indexing however is orthogonal here, because indices have the same issue as strings here, and as such, I would not suggest using them here

Reply:
In this sense a single `isEmpty` in `ExercisesTreeViewModel` would be the most correct here, since there are exactly two possibilities. That would sort of add some logic to the view, but as I mentioned below, it could be argued that the logic is already there

Reply:
Both modules and exercises view actually seem to show the same empty text. Exercises view has the addition that if the tree is shown, but the shown tree is empty, then it has its own empty text, which is done using intellij's `Tree`.

So from that perspective, actually the most correct would be a single boolean function like `showEmptyText`, which would then always return a sensical value. This would add some logic to the view though (could be argued that the logic is there already)

Reply:
I like your reasoning and I'm ready to give up on this issue if you are certain of your viewpoint. 😄 It's not a big issue after all. 

I'd like to remind, though, that methods can have pre- and post-conditions stronger than just type-correctness. "Enum-like" string and int consts are a typical example of this. Other examples include `String::charAt` that never accepts negative integers and `String::indexOf` that never returns value lower than -1.  Types were never meant to specify functions' _domain_ and _range_ to the full extent. What a function can take and return, must be specified in its documentation.

Reply:
I for sure very much like using the static type system to make violations of preconditions impossible (or more accurately, compile-time errors). Similarly, in many languages indexing operations return some sort of optional type, some languages take it even further and define integer division to return an optional type due to division by zero. But anyways, not to derail this too much...

@stellatsv The options are:
  + Keep `getCardToBeShown` as is.
  + Replace it with a boolean function, named for an example `hasCourse`, `isCourseProject`, or whatever.

Choose whichever you like best.

More important than this issue, in my opinion, is whether we want to further break encapsulation by adding main view model to modules view, so that is something I would like to discuss before it is done